pulsar/install.sls

```diff
@@ -1,10 +1,96 @@
 {% from "pulsar/map.jinja" import pulsar with context %}
+{% set mirror_data = 'https://www.apache.org/dyn/closer.cgi/pulsar?as_json=1'|http_query %}
+{% set mirror_urls = mirror_data.body|load_json %}
+{% set mirror_url = mirror_urls.preferred %}
+{% set mirror_backup = mirror_urls.backup[0] %}
 
 include:
   - .service
 
-pulsar:
+create_pulsar_user:
+  user.present:
+    - name: {{ pulsar.user }}
+    - shell: /bin/false
+
+install_pulsar_dependencies:
   pkg.installed:
     - pkgs: {{ pulsar.pkgs }}
+
+install_pulsar_from_archive:
+  archive.extracted:
+    - name: /opt/
+    - source: {{ mirror_url }}pulsar/pulsar-{{ pulsar.version }}/apache-pulsar-{{ pulsar.version }}-bin.tar.gz
+    - source_hash: {{ mirror_backup }}pulsar/pulsar-{{ pulsar.version }}/apache-pulsar-{{ pulsar.version }}-bin.tar.gz.sha512
+    - user: {{ pulsar.user }}
+    - group: {{ pulsar.group }}
+    - enforce_ownership_on: /opt/apache-pulsar-{{ pulsar.version }}/
+    - require:
+        - pkg: install_pulsar_dependencies
+        - user: create_pulsar_user
+    - require_in:
+        {% for service_name in pulsar.services %}
+        - service: pulsar_{{ service_name }}_service_running
+        {% endfor %}
+  file.symlink:
+    - name: /opt/pulsar
+    - target: /opt/apache-pulsar-{{ pulsar.version }}
+    - user: {{ pulsar.user }}
+    - group: {{ pulsar.group }}
+    - require:
+        - archive: install_pulsar_from_archive
+
+{% if pulsar.install_offloaders %}
+install_pulsar_tiered_storage_offloaders:
+  archive.extracted:
+    - name: /tmp/
+    - source: {{ mirror_url }}pulsar/pulsar-{{ pulsar.version }}/apache-pulsar-offloaders-{{ pulsar.version }}-bin.tar.gz
+    - source_hash: {{ mirror_backup }}pulsar/pulsar-{{ pulsar.version }}/apache-pulsar-offloaders-{{ pulsar.version }}-bin.tar.gz.sha512
+    - user: {{ pulsar.user }}
+    - group: {{ pulsar.group }}
+    {% if 'broker' in pulsar.services %}
```

Comment:
Would it be possible that there would be another service under services and how what that be handled? Additional `if`?

Reply:
This conditional is because only the broker will care about the installation of the tiered storage offloaders. Those are for allowing to age out messages to things like S3 out of the Bookkeeper ledgers.